data/beaconrestapi/src/main/java/tech/pegasys/teku/beaconrestapi/handlers/v2/beacon/GetFinalizedCheckpointState.java

```diff
@@ -0,0 +1,148 @@
+import static tech.pegasys.teku.infrastructure.http.RestApiConstants.RES_INTERNAL_ERROR;
+import static tech.pegasys.teku.infrastructure.http.RestApiConstants.RES_NOT_FOUND;
+import static tech.pegasys.teku.infrastructure.http.RestApiConstants.RES_OK;
+import static tech.pegasys.teku.infrastructure.http.RestApiConstants.RES_SERVICE_UNAVAILABLE;
+import static tech.pegasys.teku.infrastructure.http.RestApiConstants.SERVICE_UNAVAILABLE;
+import static tech.pegasys.teku.infrastructure.http.RestApiConstants.TAG_BEACON;
+import static tech.pegasys.teku.infrastructure.json.types.CoreTypes.BOOLEAN_TYPE;
+
+import com.fasterxml.jackson.core.JsonProcessingException;
+import io.javalin.http.Context;
+import io.javalin.plugin.openapi.annotations.HttpMethod;
+import io.javalin.plugin.openapi.annotations.OpenApi;
+import io.javalin.plugin.openapi.annotations.OpenApiContent;
+import io.javalin.plugin.openapi.annotations.OpenApiResponse;
+import java.util.Optional;
+import org.jetbrains.annotations.NotNull;
+import tech.pegasys.teku.api.ChainDataProvider;
+import tech.pegasys.teku.api.DataProvider;
+import tech.pegasys.teku.api.response.v2.beacon.GetFinalizedCheckpointStateResponse;
+import tech.pegasys.teku.api.schema.Version;
+import tech.pegasys.teku.beaconrestapi.MigratingEndpointAdapter;
+import tech.pegasys.teku.infrastructure.async.SafeFuture;
+import tech.pegasys.teku.infrastructure.json.types.SerializableTypeDefinition;
+import tech.pegasys.teku.infrastructure.restapi.endpoints.AsyncApiResponse;
+import tech.pegasys.teku.infrastructure.restapi.endpoints.EndpointMetadata;
+import tech.pegasys.teku.infrastructure.restapi.endpoints.RestApiRequest;
+import tech.pegasys.teku.spec.datastructures.metadata.ObjectAndMetaData;
+import tech.pegasys.teku.spec.datastructures.metadata.StateAndMetaData;
+import tech.pegasys.teku.spec.schemas.SchemaDefinitionCache;
+import tech.pegasys.teku.spec.schemas.SchemaDefinitions;
+
+public class GetFinalizedCheckpointState extends MigratingEndpointAdapter {
+
+  public static final String ROUTE = "/eth/v2/checkpoint/finalized_state";
+  private final ChainDataProvider chainDataProvider;
+
+  public GetFinalizedCheckpointState(
+      final DataProvider dataProvider, final SchemaDefinitionCache schemaDefinitionCache) {
+    this(dataProvider.getChainDataProvider(), schemaDefinitionCache);
+  }
+
+  public GetFinalizedCheckpointState(
+      final ChainDataProvider chainDataProvider,
+      final SchemaDefinitionCache schemaDefinitionCache) {
+    super(
+        EndpointMetadata.get(ROUTE)
+            .operationId("getFinalizedCheckpointState")
+            .summary("Get full BeaconState object for finalized checkpoint state")
+            .description(
+                "Returns full BeaconState object for a finalized checkpoint state from the WS period. \n"
+                    + "Depending on `Accept` header it can be returned either as json or as bytes serialized by SSZ.")
+            .tags(TAG_BEACON)
```

Comment:
We should add the experimental tag here as well.  @rolfyone What else should we do to make it clear this API may be changed or disappear without warning?

Reply:
> Looking good I think. The main thing will just be making it clear that this API isn't yet stable and may change. In particular there's open questions about whether it should support JSON at all and what it should do if the state is only optimistically synced.

Yep, for now I am keeping it as a draft PR while waiting for more clarifications... I'll adjust according to the ongoing discussion.

Reply:
tagging as experimental without the beacon flag if we're really unsure of it's future... if we think its probably going to be pretty much right I'd leave beacon flag on it, but typically i only flag experimental until PRs are approved in beacon-apis...

Reply:
The beacon-apis PR is not merged yet so I tagged it experimental.
Will update it to beacon when the PR is merged